File: server/Subscription.DomainService/Utilities/MeterQuantity.cs
using System.Globalization;

namespace Subscription.DomainService.Utilities;

/// <summary>
/// The rules a metered quantity obeys, and the one place a fractional charge becomes money.
/// </summary>
/// <remarks>
/// Quantities are <see cref="decimal"/> — exact base-ten, stored as BSON <c>Decimal128</c> — rather
/// than <see cref="double"/>. A reversal has to cancel the entry it compensates to the last place,
/// and binary floating point cannot promise that: <c>0.1 + 0.2 - 0.3</c> leaves a residue that would
/// sit in a customer's balance forever.
/// <para>
/// Fractions are opt-in per meter. A meter's <c>QuantityScale</c> is how many decimal places it
/// accepts, and it defaults to zero — so a meter that counts screenings keeps refusing half of one,
/// exactly as it did when these fields were integers. A plan already in the database has no such
/// field, deserializes to zero, and therefore cannot behave differently.
/// </para>
/// </remarks>
public static class MeterQuantity
{
    /// <summary>
    /// The finest granularity any meter may declare.
    /// </summary>
    /// <remarks>
    /// Six places covers the units that prompted this — storage in GB, time in hours, prorated
    /// credits — and keeps every quantity below roughly nine billion exactly representable as an
    /// IEEE-754 double, so a browser renders the figure the server computed rather than one that
    /// drifted in transit.
    /// </remarks>
    public const int MaxScale = 6;

    /// <summary>The largest magnitude a quantity may take.</summary>
    /// <remarks>
    /// Decimal128 holds more than <see cref="decimal"/> does, so a value written by something other
    /// than this service could otherwise overflow on the way back in. Bounding it at authoring and
    /// recording time means that never happens. Generous enough that no real meter reaches it, and
    /// it still leaves headroom for the tier multiplication below.
    /// </remarks>
    public const decimal MaxMagnitude = 1_000_000_000_000m;

    /// <summary>How many decimal places a value actually carries, trailing zeroes ignored.</summary>
    /// <remarks>
    /// Read from the decimal's own scale after normalising, because <c>1.50m</c> and <c>1.5m</c> are
    /// equal but differently scaled — refusing the first on a one-place meter would be arbitrary.
    /// </remarks>
    public static int ScaleOf(decimal value)
    {
        var scale = (decimal.GetBits(value)[3] >> 16) & 0xFF;

        // Rounding to one place fewer leaves the value untouched exactly when the last place holds
        // a zero, so this walks off the trailing zeroes without any string formatting.
        while (scale > 0 && value == decimal.Round(value, scale - 1))
        {
            scale--;
        }

        return scale;
    }

    /// <summary>Whether a meter declaring <paramref name="scale"/> may hold this quantity.</summary>
    public static bool IsWithinScale(decimal value, int scale) =>
        scale >= 0 &&
        scale <= MaxScale &&
        ScaleOf(value) <= scale;

    /// <summary>Whether a quantity is inside the representable range.</summary>
    public static bool IsWithinMagnitude(decimal value) =>
        Math.Abs(value) <= MaxMagnitude;

    /// <summary>Whether a scale is one a meter may declare at all.</summary>
    public static bool IsValidScale(int scale) =>
        scale is >= 0 and <= MaxScale;

    /// <summary>
    /// Turns an exactly-computed charge into whole minor units.
    /// </summary>
    /// <remarks>
    /// The single rounding event in usage rating. Tier arithmetic stays exact all the way through —
    /// so re-banding a rate table without changing any of its prices cannot change the bill — and
    /// only the meter's own total is rounded, because that total is the invoice line the customer
    /// actually sees.
    /// <para>
    /// Away from zero at the midpoint: half a minor unit rounds up, which is the convention an
    /// invoice reader expects and which a reversal mirrors symmetrically, since the sign is carried
    /// through rather than truncated toward it.
    /// </para>
    /// </remarks>
    public static long ToMinorUnits(decimal exactAmountMinor)
    {
        var rounded = decimal.Round(exactAmountMinor, 0, MidpointRounding.AwayFromZero);

        // Checked, for the same reason the tier walk is: a technically valid rate and a technically
        // valid quantity can each pass validation alone and still multiply into something no
        // long-minor amount can hold. Refusing beats silently wrapping the charge.
        return checked((long)rounded);
    }

    /// <summary>
    /// The smallest quantity a meter at this scale can distinguish: <c>1</c> at scale zero,
    /// <c>0.001</c> at scale three.
    /// </summary>
    /// <remarks>
    /// Used to name the first quantity inside a rate band whose lower bound is exclusive. At scale
    /// zero that lands on the whole unit the band has always started at, so a plan authored before
    /// fractions existed reports exactly the bands it reported before.
    /// </remarks>
    public static decimal SmallestStep(int scale) =>
        scale switch
        {
            <= 0 => 1m,
            1 => 0.1m,
            2 => 0.01m,
            3 => 0.001m,
            4 => 0.0001m,
            5 => 0.00001m,
            _ => 0.000001m
        };

    /// <summary>A quantity as it should appear in a log line or an error message.</summary>
    public static string Describe(decimal value) =>
        value.ToString("0.######", CultureInfo.InvariantCulture);
}

File: client/app/cross-modules/utilities/subscription/utilities/meter-quantity.ts
/**
 * The granularity rules a metered quantity obeys, mirroring the server's `MeterQuantity`.
 *
 * Fractions are opt-in per meter: a meter's `quantityScale` is how many decimal places it accepts,
 * and zero — the default, and what a meter authored before fractions existed reports — means whole
 * units only. Checking it here is so the author is told in the form rather than by a rejected
 * request; the server refuses the same values regardless of what happens in the browser.
 */

/** The finest granularity any meter may declare. Must match `MeterQuantity.MaxScale`. */
export const METER_QUANTITY_MAX_SCALE = 6;

/** The largest magnitude a quantity may take. Must match `MeterQuantity.MaxMagnitude`. */
export const METER_QUANTITY_MAX_MAGNITUDE = 1_000_000_000_000;

/**
 * How many decimal places a number actually carries.
 *
 * Read off the decimal string rather than by arithmetic, because scaling by a power of ten to
 * "check for a remainder" is exactly the kind of binary floating-point step that reports 1.15 as
 * having a remainder. Exponent form is handled because that is how JavaScript renders anything
 * below 1e-6, which is the first thing a six-place meter can hold.
 */
export const scaleOf = (value: number): number => {
  if (!Number.isFinite(value)) {
    return Number.POSITIVE_INFINITY;
  }

  const text = Math.abs(value).toString();
  const exponentAt = text.indexOf("e");

  if (exponentAt >= 0) {
    const mantissa = text.slice(0, exponentAt);
    const power = Number(text.slice(exponentAt + 1));
    const fraction = mantissa.split(".")[1] ?? "";

    return Math.max(0, fraction.length - power);
  }

  const pointAt = text.indexOf(".");

  return pointAt < 0 ? 0 : text.length - pointAt - 1;
};

/** Whether a meter declaring this scale may hold this quantity. */
export const isWithinScale = (value: number, scale: number): boolean =>
  Number.isFinite(value) && scaleOf(value) <= scale;

/** Whether a quantity is inside the representable range. */
export const isWithinMagnitude = (value: number): boolean =>
  Number.isFinite(value) && Math.abs(value) <= METER_QUANTITY_MAX_MAGNITUDE;

/**
 * The `step` a number input should take for a meter at this scale.
 *
 * Without it the browser's own validation rejects a fraction on a `type="number"` field whose step
 * defaults to one, before any of the above is ever consulted.
 */
export const stepFor = (scale: number): string =>
  scale <= 0 ? "1" : (1 / 10 ** scale).toFixed(scale);

/**
 * A quantity as it should be displayed: no trailing zeroes, and no exponent form.
 *
 * A figure that came back from the server's Decimal128 arithmetic can carry places the author
 * never typed, and an allowance of five hundred should not read as `500.000000`.
 */
export const formatQuantity = (value: number): string =>
  Number.isFinite(value)
    ? value.toLocaleString(undefined, {
        minimumFractionDigits: 0,
        maximumFractionDigits: METER_QUANTITY_MAX_SCALE,
      })
    : "";
